### DISCUSSION TITLE
Starchart: publiccodeyml deficiencies

### DISCUSSION THREAD

**Discussion by realaravinth** [publiccodeyml/publiccode.yml#157] [Ideas]
> @ruphy kindly [prompted](https://github.com/forgeflux-org/starchart/issues/3#issuecomment-1118186023) me to start a thread on here

Hello, I'm from the [Forgeflux project](https://forgeflux.org)(software forge federation) currently involved in writing a software forge spider for crawling and indexing public, Free(as in Freedom) software repositories hosted on independent forge instances. We have plans to implement official support for Gitea(WIP), GitLab and SourceHut.

Starchart will use forge-provided APIs(like the [Gitea API](https://try.gitea.io/api/swagger)) to request and store information on the publicly available repositories hosted on the instance. It will also periodically crawl indexed instances to update the data. The goal
is to create a Free, federated alternative to third-party indexes like awesomeopensource.com to improve visibility for projects hosted on
independently run forges.

Each Starchart repository, in addition to storing data in a database, will store indexed data in a Git repository. This Git repository will be
publicly available and will be used as the primary federation mechanism: federating Starchart instances will `git pull` each other's indexed data.

[@6543 kindly referred me to this project](https://github.com/forgeflux-org/starchart/issues/3#issuecomment-1118186023) for use as Starchart's file storage format. publiccodeyml requires(mandatory fields) data that can't be retrieved/automatically inferred from data that is provided by the forges' APIs. The following is a list of mandatory fields that can't be supported by Starchart:

1. [`platforms`](https://yml.publiccode.tools/schema.core.html#key-platforms) describes the platform the software will run on, which is hard to determine automatically by Starchart.
2. [`categories`](https://yml.publiccode.tools/schema.core.html#key-categories): defined as the list of words used to describe the software. It may be derived from repository tags, but tags on forges are optional so Starchart will be breaking the mandatory requirement.
3. [`developmentStatus`](https://yml.publiccode.tools/schema.core.html#key-developmentstatus): `concept`, `development`, `beta`, etc. (see full list in link), are not possible to determine.
4. [`softwareType`](https://yml.publiccode.tools/schema.core.html#key-softwaretype) is similar to `platform` but is more fine-grained, again, something that Starchart can't auto-generate.
5. [`description/[lang]/genericName`](https://yml.publiccode.tools/schema.core.html#key-description-lang-genericname)
6. [`description/[lang]/longDescription`](https://yml.publiccode.tools/schema.core.html#key-description-lang-longdescription) may be derived from repository description, but it is an optional field in forges
7. [`description/[lang]/features`](https://yml.publiccode.tools/schema.core.html#key-description-lang-features) mandatory per schema but not possible to auto generate
8. [`legal/licence`](https://yml.publiccode.tools/schema.core.html#key-legal-license) hard to generate when repository doesn't contain a licence file
9. [`maintainance/type`](https://yml.publiccode.tools/schema.core.html#key-maintenance-type) is hard to determine
10. [`maintenance/contractors`](https://yml.publiccode.tools/schema.core.html#key-maintenance-contractors): conditionally mandatory if `maintainance/type` is `contract` and is impossible for the same reasons as `maintainance/type`.
11. [`maintainance/contacts`](https://yml.publiccode.tools/schema.core.html#key-maintenance-contacts): can be generated from the repository owner's email address but isn't ethical to re-share when the owner hasn't explicitly authorized Starchart to share their email.
12. [`localisation/localisationReady`](https://yml.publiccode.tools/schema.core.html#key-localisation-localisationready)
13. [`localisation/availableLanguages`](https://yml.publiccode.tools/schema.core.html#key-localisation-availablelanguages)

## Solution

I propose the following solution, but I'm open for suggestions:

1. Support fields that are possible fields that can be retrieved.
2. Check for the existence of a `publiccode.yml` in the repository at the file locations mandated by the publiccodeyml standard.
3. If `publiccode.yml` exists, store a link to the file in the database and the repository information file.

Thoughts? :)

**Comment by realaravinth** [comment]
I'm happy to report that partial support for `publiccodeyml`, as proposed above, is implemented in starchart! :tada: 

The first instance of starchart is online at [starchart.forgeflux.org](https://starchart.forgeflux.org) with support for only Gitea at the moment(support for GitLab and Sourcehut will be added soon).

The crawled data is shared publicly from day 0: to download data, head over to [/federate](https://starchart.forgeflux.org/federate/) and grab the latest tarball. Please see [here](https://github.com/forgeflux-org/starchart/blob/master/docs/published-crwaled-data.md) for more information about the format and structure of the published data. The format may change before starchart stabilizes, but I will keep the linked document updated.

As always, I'm open for any suggestions that you might have :)

Cheers!

**Reply by bzg** [reply]
Congrats! What is the license for published data? I suggest [CC0](https://creativecommons.org/share-your-work/public-domain/cc0/) but maybe you have some constraints.

**Reply by realaravinth** [reply]
Thanks!

I didn't think about the license because starchart is simply republishing data that is available in public domain, but I'll explicitly mention that the information is in public domain :)

**Reply by realaravinth** [reply]
Done! CC0 is [included](https://github.com/forgeflux-org/starchart/commit/036a90e74ac292b46738e15e3c562e9406915f4d) with [published data](https://starchart.forgeflux.org/federate/).

**Comment by bzg** [comment]
Great, thanks!
